docs: add example systemd service file

sangaman opened the pull request:
This is an example systemd service file for lnd designed to run alongside a bitcoind service. This file is loosely based on https://github.com/bitcoin/bitcoin/blob/master/contrib/init/bitcoind.service, https://gist.github.com/Stadicus/f0c6db4fa6cf787f19d9d3444b91633f, and my own experience managing lnd with systemd.

I'm not an expert on systemd by any means so I'm open to suggestions, but right now the only example lnd service files I'm aware of on the internet are flawed and/or stale (see https://github.com/lightningnetwork/lnd/issues/5996#issuecomment-976052716). I also wasn't sure whether this is the sort of change that merits a mention in the release notes, so I left that out.

guggero commented on contrib/init/lnd.service:
```diff
@@ -0,0 +1,36 @@
+# A sample systemd service file for lnd running with a bitcoind service.
+
+[Unit]
+Description=LND Lightning Daemon
```
nit: lnd stands for lightning network daemon

guggero reviewed (approved):
Nice addition, thank you! Just needs an entry in the release notes (or the `[skip ci]` suffix in the commit message), otherwise LGTM :tada:

sangaman commented:
@guggero Thanks for the quick review. I updated the service `Description` per your comment and added a release notes file with a short mention of the addition for 0.15.0 since one does not yet exist - I'm assuming that's the next targeted release.

guggero commented on docs/release-notes/release-notes-0.15.0.md:
```diff
@@ -0,0 +1,10 @@
+
+## Misc
+
+* An example systemd service file for running lnd alongside a bitcoind service
```
To satisfy the CI check, you need to include a link to the PR. You can either link the whole sentence or maybe just the `example systemd service file`.

sangaman replied:
Oops, thank you. Done.

As a side note it seems a bit self-referential in that you need the PR number to pass the check, but you don't know the number until you open the PR (with a failing release notes check, presumably).

guggero replied:
Yes, that's true. Though for the release notes it is nice to link to the PR that added the change so you can easily look up all the context if you're interested. So unless you're guessing the next PR number from looking at the last created issue/PR you need to push at least a second time, unfortunately.

theborakompanioni commented:
@sangaman Just a question: Why not also add additional "hardening measures" as per the bitcoin systemd config example (e.g. `MemoryDenyWriteExecute`, `PrivateDevices`, `PrivateTmp`, etc.)?

bhandras commented on contrib/init/README.md:
```diff
@@ -0,0 +1,5 @@
+Sample configuration files for:
```
nit: it could be also useful to detail how to enable/start/stop.

sangaman replied:
Something like this?

```bash
# Enable lnd to automatically start on system boot
systemctl enable lnd

# Start lnd
systemctl start lnd

# Restart lnd
systemctl restart lnd

# Stop lnd
systemctl stop lnd
```

bhandras reviewed (approved):
LGTM, thanks @sangaman :+1:

theborakompanioni commented:
> So I can add `PrivateDevices` and maybe `MemoryDenyWriteExecute` as well to the hardening section if nobody has any other concerns.

While `ProtectHome=true` might not work in this setup, lnd is running here without issues with the following settings (same as [bitcoind.service](https://github.com/bitcoin/bitcoin/blob/master/contrib/init/bitcoind.service)):
```
# Provide a private /tmp and /var/tmp.
PrivateTmp=true

# Mount /usr, /boot/ and /etc read-only for the process.
ProtectSystem=full

# Deny access to /home, /root and /run/user
ProtectHome=true

# Disallow the process and all of its children to gain
# new privileges through execve().
NoNewPrivileges=true

# Use a new /dev namespace only populated with API pseudo devices
# such as /dev/null, /dev/zero and /dev/random.
PrivateDevices=true

# Deny the creation of writable and executable memory mappings.
MemoryDenyWriteExecute=true
```

Just in case you were wondering whether they work at all. Maybe it is not necessary, especially when it should just be an example service file. Users will adapt to their needs anyway. Keep it simple :+1:

sangaman commented:
I updated the PR to add the hardening suggestions above (except `PrivateTmp` since that would conflict with my setup and potentially others) and also to add some example commands to the init README as suggested by @bhandras.

Kixunil commented on contrib/init/lnd.service:
```diff
@@ -0,0 +1,50 @@
+####################
+
+# Mount /usr, /boot/ and /etc read-only for the process.
+ProtectSystem=full
```
FYI, `ProtectHome=true` should also work as long as you set `ReadWriteDirectories=/home/bitcoin/.lnd/`

Kixunil commented on contrib/init/lnd.service:
```diff
@@ -0,0 +1,50 @@
+Group=bitcoin
+
+# Try restarting lnd if it stops due to a failure
+Restart=on-failure
```
I'd recommend using `always` in case `lnd` accidentally exits with code 0. It can still be stopped by explicit `systemctl` command, so no harm there and most people probably want their channels to be watched as reliably as possible.

sangaman replied:
Would it restart automatically if you used `lncli stop` with `always`?

Kixunil replied:
No, `systemctl stop` will keep it stopped which I believe is good. One could add `RefuseManualStop` to avoid stopping it accidentally but IMO that's too much.

sangaman replied:
Yes for `systemctl stop` but what about `lncli stop` or otherwise shutting down the node via the gRPC layer? I figure ideally both should actually stop it  but I assumed the latter would not stop it for good with `Restart=always`.

Kixunil replied:
Sorry, misread. Yes, `lncli stop` wouldn't work. I personally don't like stopping thing via RPC but I can understand that people would find it surprising to not work, so not pushing it.

Kixunil commented on contrib/init/README.md:
````diff
@@ -0,0 +1,27 @@
+systemctl stop lnd
+```
+
+Systemd will attempt to restart lnd automatically if it crashes or otherwise stops unexpectedly.
````
Could also include a note about configuring services that depend on LND (e.g. ThunderHub) with the `After=lnd.service` option.

Kixunil reviewed (commented):
Sorry for being late, found this randomly and got interested. I had some more ideas to improve.

@sangaman what do you mean by using `/tmp`  for zmq block and tx data? Do you use ZMQ with Unix socket which is bound there?

sangaman commented:
> Sorry for being late, found this randomly and got interested. I had some more ideas to improve.

Thanks for the suggestions, they'd be helpful for others who might come across this or maybe for a new PR.

> @sangaman what do you mean by using `/tmp` for zmq block and tx data? Do you use ZMQ with Unix socket which is bound there?

Yes, I'm using the `/tmp` folder for the unix socket for the zmq notifications. That's just based off an example I came across a long time ago, and it has worked fine for me so I refrained from changing it.

Kixunil commented:
In that case I recommend putting it to `/var/run/yourdirectory` in systemd which is the usual location for such things in Linux and you can control access to it better. `/tmp` is not intended for such things. If you decide to do so use `RuntimeDirectory=yourdirectory` in systemd

sangaman commented:
> In that case I recommend putting it to `/var/run/yourdirectory` in systemd which is the usual location for such things in Linux and you can control access to it better. `/tmp` is not intended for such things. If you decide to do so use `RuntimeDirectory=yourdirectory` in systemd

I'll try `/var` next time I restart my node. Thanks for the tip.
